espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/state/InputStateTable.java

```diff
@@ -0,0 +1,112 @@
+package io.appium.espressoserver.lib.helpers.w3c.state;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import io.appium.espressoserver.lib.handlers.exceptions.AppiumException;
+import io.appium.espressoserver.lib.helpers.w3c.adapter.W3CActionAdapter;
+import io.appium.espressoserver.lib.helpers.w3c.models.ActionObject;
+
+import io.appium.espressoserver.lib.helpers.w3c.models.ActionObject;
+
+/**
+ * Keep the state of all active input sources
+ *
+ * (defined in 17.1 of spec)
+ */
+public class InputStateTable {
+
+    private final Map<String, InputState> stateTable = new HashMap<>();
+    private final List<ActionObject> cancelList = new ArrayList<>();
+
+    private static Map<String, InputStateTable> inputStateTables = new HashMap<>();
+
+    public void addInputState(String id, InputState inputState){
+        stateTable.put(id, inputState);
+    }
+
+    public InputState getInputState(String id) {
+        return stateTable.get(id);
+    }
+
+    public InputState getOrCreateInputState(String sourceId, ActionObject actionObject) {
+        if (!this.hasInputState(sourceId)) {
+            InputState newInputState = null;
+            switch(actionObject.getType()) {
+                case KEY:
+                    newInputState = new KeyInputState();
+                    break;
+                case POINTER:
+                    newInputState = new PointerInputState();
+                    break;
+                // Don't need to track state of null input types
+                case NONE:
```

Comment:
this can still contain pauses

Reply:
Returning `null` covers `pause`.

If `InputState == null` it just does a pause.